reference/reference.go

```diff
@@ -56,6 +65,11 @@ var (
 type Reference interface {
 	// String returns the full reference
 	String() string
+
+	// internal requires that all instances of Reference must be implemented
+	// by this package. This allows us to ensure that instances can be treated
+	// as an algebraic data type.
+	internal()
```

Comment:
I understand that you want to force the requirement, but I think it needs to come *after* the package is unified. I don't want to merge something that will make it impossible to vendor an updated distribution package without further big changes, especially when I need to vendor an updated distribution package soon.